File: backend-gin/internal/handlers/patients.go
package handlers

import (
	"net/http"
	"strconv"
	"strings"
	"time"

	"backend-gin/internal/database" // Assuming this path is correct based on previous steps
	"backend-gin/internal/models"   // Assuming this path is correct
	"github.com/gin-gonic/gin"
	"gorm.io/gorm"
)

// --- Structs for Request Binding ---

type CreatePatientRequest struct {
	Username string `json:"username" binding:"required"`
	Age      int    `json:"age" binding:"required"` // Model has *int, so &req.Age is fine
	Gender   string `json:"gender" binding:"required"` // Model has *string, so &req.Gender is fine
	CaseID   string `json:"case_id" binding:"required"`
	DoctorID uint   `json:"doctor_id" binding:"required"` // Model has *uint, so &req.DoctorID is fine
}

type UpdatePatientRequest struct {
	PatientID uint   `json:"patient_id" binding:"required"`
	Username  string `json:"username" binding:"required"`
	Age       int    `json:"age" binding:"required"`
	Gender    string `json:"gender" binding:"required"`
	CaseID    string `json:"case_id" binding:"required"`
	DoctorID  uint   `json:"doctor_id" binding:"required"` // This is the ID of the doctor performing the update
}

type PatientLoginRequest struct {
	CaseID       string `json:"case_id" binding:"required"`
	VerifyCaseID string `json:"verify_case_id" binding:"required"`
}

// --- Handler Functions ---

func PatientLogin(c *gin.Context) {
	var req PatientLoginRequest
	if err := c.ShouldBindJSON(&req); err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"error": err.Error()})
		return
	}

	var patient models.Patient
	// Query for patient by CaseID where IsDeleted is false
	if err := database.DB.Where("case_id = ? AND is_deleted = ?", req.CaseID, false).First(&patient).Error; err != nil {
		if err == gorm.ErrRecordNotFound {
			c.JSON(http.StatusNotFound, gin.H{"message": "Patient not found"})
			return
		}
		c.JSON(http.StatusInternalServerError, gin.H{"message": "Database error", "details": err.Error()})
		return
	}

	// The original FastAPI code checks if patient.case_id == req.verify_case_id.
	// If CaseID is unique and we fetch based on req.CaseID, this check is primarily to ensure req.VerifyCaseID is also correct.
	if patient.CaseID != req.VerifyCaseID {
		c.JSON(http.StatusBadRequest, gin.H{"message": "Case ID verification failed. Ensure 'case_id' and 'verify_case_id' match and are correct."})
		return
	}

	c.JSON(http.StatusOK, gin.H{"message": "Login successful", "patient": patient})
}

func InsertPatient(c *gin.Context) {
	var req CreatePatientRequest
	if err := c.ShouldBindJSON(&req); err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"error": err.Error()})
		return
	}

	currentTime := time.Now().Format("2006-01-02 15:04:05")
	newPatient := models.Patient{
		Username:   req.Username,
		Age:        &req.Age,    // models.Patient.Age is *int
		Gender:     &req.Gender, // models.Patient.Gender is *string
		CaseID:     req.CaseID,
		DoctorID:   &req.DoctorID, // models.Patient.DoctorID is *uint
		CreateTime: currentTime,
		UpdateTime: currentTime,
		IsDeleted:  false,
	}

	if err := database.DB.Create(&newPatient).Error; err != nil {
		// Check for unique constraint violation on CaseID if your DB schema has it
		if strings.Contains(err.Error(), "UNIQUE constraint failed") || strings.Contains(err.Error(), "duplicate key value violates unique constraint") {
             c.JSON(http.StatusConflict, gin.H{"message": "Patient with this Case ID already exists", "details": err.Error()})
             return
        }
		c.JSON(http.StatusInternalServerError, gin.H{"message": "Failed to insert patient", "details": err.Error()})
		return
	}
	c.JSON(http.StatusOK, newPatient)
}

func GetAllPatientsByDoctorID(c *gin.Context) {
	doctorIDStr := c.Param("doctor_id")
	doctorID, err := strconv.ParseUint(doctorIDStr, 10, 32)
	if err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"message": "Invalid doctor ID format"})
		return
	}

	var doctor models.Doctor
	if err := database.DB.Where("id = ? AND is_deleted = ?", uint(doctorID), false).First(&doctor).Error; err != nil {
		if err == gorm.ErrRecordNotFound {
			c.JSON(http.StatusNotFound, gin.H{"message": "Doctor not found"})
			return
		}
		c.JSON(http.StatusInternalServerError, gin.H{"message": "Database error fetching doctor", "details": err.Error()})
		return
	}

	var patients []models.Patient
	query := database.DB.Where("is_deleted = ?", false)

	// RoleID in Doctor model is *uint. Need to dereference it carefully.
	// Assuming RoleID 1 is Admin. If RoleID is nil or not 1, filter by doctor_id.
	isAdmin := false
	if doctor.RoleID != nil && *doctor.RoleID == 1 {
		isAdmin = true
	}

	if !isAdmin {
		query = query.Where("doctor_id = ?", uint(doctorID))
	}

	if err := query.Find(&patients).Error; err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"message": "Database error fetching patients", "details": err.Error()})
		return
	}
	c.JSON(http.StatusOK, patients)
}

func UpdatePatientByID(c *gin.Context) {
	var req UpdatePatientRequest
	if err := c.ShouldBindJSON(&req); err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"error": err.Error()})
		return
	}

	// Verify the doctor performing the update exists
	var performingDoctor models.Doctor
	if err := database.DB.Where("id = ? AND is_deleted = ?", req.DoctorID, false).First(&performingDoctor).Error; err != nil {
		if err == gorm.ErrRecordNotFound {
			c.JSON(http.StatusNotFound, gin.H{"message": "Performing doctor not found"})
			return
		}
		c.JSON(http.StatusInternalServerError, gin.H{"message": "Database error verifying doctor", "details": err.Error()})
		return
	}

	var patientToUpdate models.Patient
	query := database.DB.Where("id = ? AND is_deleted = ?", req.PatientID, false)

	// Authorization: If the performing doctor is not an admin (RoleID != 1),
	// they can only update patients assigned to them.
	isPerformingDoctorAdmin := false
	if performingDoctor.RoleID != nil && *performingDoctor.RoleID == 1 {
		isPerformingDoctorAdmin = true
	}

	if !isPerformingDoctorAdmin {
		// Patient's DoctorID must match the performing doctor's ID
		query = query.Where("doctor_id = ?", performingDoctor.ID)
	}

	if err := query.First(&patientToUpdate).Error; err != nil {
		if err == gorm.ErrRecordNotFound {
			c.JSON(http.StatusNotFound, gin.H{"message": "Patient not found or doctor lacks permission to update"})
			return
		}
		c.JSON(http.StatusInternalServerError, gin.H{"message": "Database error finding patient", "details": err.Error()})
		return
	}

	// Update fields
	patientToUpdate.Username = req.Username
	patientToUpdate.Age = &req.Age
	patientToUpdate.Gender = &req.Gender
	patientToUpdate.CaseID = req.CaseID // Potentially update CaseID
	// The DoctorID in UpdatePatientRequest (req.DoctorID) is the ID of the doctor performing the update.
	// To change the patient's assigned doctor, we'd need another field like `TargetDoctorID` or similar.
	// For now, assuming req.DoctorID is for authorization and we don't change patientToUpdate.DoctorID unless specified.
	// If the intent is to allow changing the patient's assigned doctor, then:
	// patientToUpdate.DoctorID = &req.DoctorIDToAssign (assuming such a field exists in request)
	// For this implementation, we will NOT update patientToUpdate.DoctorID from req.DoctorID to prevent accidental reassignment
	// If re-assignment is a feature, it should be explicit.

	patientToUpdate.UpdateTime = time.Now().Format("2006-01-02 15:04:05")

	if err := database.DB.Save(&patientToUpdate).Error; err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"message": "Failed to update patient", "details": err.Error()})
		return
	}
	c.JSON(http.StatusOK, patientToUpdate)
}

func DeletePatientByID(c *gin.Context) {
	patientIDStr := c.Param("patient_id")
	patientID, errP := strconv.ParseUint(patientIDStr, 10, 32)
	if errP != nil {
		c.JSON(http.StatusBadRequest, gin.H{"message": "Invalid patient ID format"})
		return
	}

	doctorIDStr := c.Param("doctor_id") // This is the ID of the doctor performing the deletion
	performingDoctorID, errD := strconv.ParseUint(doctorIDStr, 10, 32)
	if errD != nil {
		c.JSON(http.StatusBadRequest, gin.H{"message": "Invalid doctor ID format for performing doctor"})
		return
	}

	var performingDoctor models.Doctor
	if err := database.DB.Where("id = ? AND is_deleted = ?", uint(performingDoctorID), false).First(&performingDoctor).Error; err != nil {
		if err == gorm.ErrRecordNotFound {
			c.JSON(http.StatusNotFound, gin.H{"message": "Performing doctor not found"})
			return
		}
		c.JSON(http.StatusInternalServerError, gin.H{"message": "Database error verifying doctor", "details": err.Error()})
		return
	}

	var patientToDelete models.Patient
	query := database.DB.Where("id = ? AND is_deleted = ?", uint(patientID), false)

	isPerformingDoctorAdmin := false
	if performingDoctor.RoleID != nil && *performingDoctor.RoleID == 1 {
		isPerformingDoctorAdmin = true
	}

	if !isPerformingDoctorAdmin {
		query = query.Where("doctor_id = ?", performingDoctor.ID)
	}

	if err := query.First(&patientToDelete).Error; err != nil {
		if err == gorm.ErrRecordNotFound {
			c.JSON(http.StatusNotFound, gin.H{"message": "Patient not found or doctor lacks permission for deletion"})
			return
		}
		c.JSON(http.StatusInternalServerError, gin.H{"message": "Database error finding patient for deletion", "details": err.Error()})
		return
	}

	patientToDelete.IsDeleted = true
	patientToDelete.UpdateTime = time.Now().Format("2006-01-02 15:04:05")

	if err := database.DB.Save(&patientToDelete).Error; err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"message": "Failed to delete patient", "details": err.Error()})
		return
	}
	c.JSON(http.StatusOK, gin.H{"message": "Patient deleted successfully"})
}

func GetPatientsWithPage(c *gin.Context) {
	pageStr := c.DefaultQuery("page", "1")
	pageSizeStr := c.DefaultQuery("page_size", "10")
	sortBy := c.DefaultQuery("sort_by", "id") // Field name from models.Patient
	sortOrder := strings.ToLower(c.DefaultQuery("sort_order", "asc"))
	doctorIDStr := c.Query("doctor_id") // Optional filter

	page, err := strconv.Atoi(pageStr)
	if err != nil || page < 1 {
		page = 1
	}
	pageSize, err := strconv.Atoi(pageSizeStr)
	if err != nil || pageSize < 1 {
		pageSize = 10
	}

	query := database.DB.Model(&models.Patient{}).Where("is_deleted = ?", false)

	if doctorIDStr != "" {
		doctorID, err := strconv.ParseUint(doctorIDStr, 10, 32)
		if err == nil {
			query = query.Where("doctor_id = ?", uint(doctorID))
		} else {
			c.JSON(http.StatusBadRequest, gin.H{"message": "Invalid doctor_id format for filtering"})
			return
		}
	}

	var total int64
	if err := query.Count(&total).Error; err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"message": "Error counting patients", "details": err.Error()})
		return
	}

	// Validate sortBy to prevent SQL injection and map to actual DB column names if necessary
	// For simplicity, assuming direct mapping for id, username, create_time.
	// A more robust solution uses a map of allowed sort fields.
	allowedSortFields := map[string]string{
		"id":         "id",
		"username":   "username",
		"case_id":    "case_id",
		"age":        "age",
		"gender":     "gender",
		"doctor_id":  "doctor_id",
		"createtime": "create_time", // frontend might send createtime
		"create_time": "create_time",
		"updatetime": "update_time", // frontend might send updatetime
		"update_time": "update_time",

	}
	dbSortField, isValidSortField := allowedSortFields[strings.ToLower(sortBy)]
	if !isValidSortField {
		dbSortField = "id" // Default sort field
	}

	if sortOrder != "asc" && sortOrder != "desc" {
		sortOrder = "asc"
	}
	orderClause := dbSortField + " " + sortOrder
	query = query.Order(orderClause)

	offset := (page - 1) * pageSize
	var patients []models.Patient
	if err := query.Offset(offset).Limit(pageSize).Find(&patients).Error; err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"message": "Error fetching paginated patients", "details": err.Error()})
		return
	}

	c.JSON(http.StatusOK, gin.H{
		"total":     total,
		"page":      page,
		"page_size": pageSize,
		"patients":  patients,
	})
}

File: backend/apis/patients.py
from datetime import datetime
from typing import Optional

from fastapi import APIRouter, Body, Query
from models import Doctors, Patients, SessionDep, VideoPath, Action
from pydantic import BaseModel
from sqlalchemy import asc, desc, func, select, update

router = APIRouter(tags=["patients"], prefix="/patients")


class CreatePatientModel(BaseModel):
    username: str
    age: int
    gender: str
    case_id: str
    patient_id: int
    doctor_id: int


class UpdatePatientModel(BaseModel):
    age: int
    gender: str
    case_id: str
    doctor_id: int
    patient_id: int  # 当前患者ID，用于查找患者
    new_patient_id: int  # 新的患者ID（病历号）
    username: str


class PatientLoginModel(BaseModel):
    case_id: str
    verify_case_id: str


@router.post("/patient_login")
async def patient_login(patient: PatientLoginModel = Body(..., embed=True), session: SessionDep = SessionDep):
    result = await session.execute(select(Patients).where(
        Patients.case_id == patient.case_id, Patients.is_deleted == False))
    if patient_ := result.scalar_one_or_none():
        return (
            {"message": "Case ID verification failed"}
            if patient_.case_id != patient.verify_case_id
            else {"message": "Login successful", "patient": patient_.to_dict()}
        )
    else:
        return {"message": "Patient not found"}


@router.put("/insert_patient")
async def insert_patient(patient: CreatePatientModel = Body(..., embed=True), session: SessionDep = SessionDep):
    result = await session.execute(select(Doctors).where(
        Doctors.id == patient.doctor_id, Doctors.is_deleted == False))
    doctor = result.scalar_one_or_none()
    if not doctor:
        return {"message": "Doctor not found"}

    # 检查case_id是否已存在
    result = await session.execute(select(Patients).where(
        Patients.case_id == patient.case_id, Patients.is_deleted == False))
    if patient_ := result.scalar_one_or_none():
        return {"message": "Case ID already exists"}

    # 检查patient_id是否已存在
    result = await session.execute(select(Patients).where(
        Patients.id == patient.patient_id, Patients.is_deleted == False))
    if patient_ := result.scalar_one_or_none():
        return {"message": "Patient ID already exists"}

    patient_obj = Patients(username=patient.username, age=patient.age, gender=patient.gender, case_id=patient.case_id, doctor_id=patient.doctor_id,
                           create_time=datetime.now().strftime("%Y-%m-%d %H:%M:%S"), update_time=datetime.now().strftime("%Y-%m-%d %H:%M:%S"), is_deleted=False)
    # 手动设置ID
    patient_obj.id = patient.patient_id
    session.add(patient_obj)
    await session.commit()
    # Refresh to get the ID and avoid detached instance
    await session.refresh(patient_obj)
    return patient_obj.to_dict()


@router.get("/get_all_patient_by_doctor_id/{doctor_id}")
async def get_all_patient_by_doctor_id(doctor_id: int, session: SessionDep = SessionDep):
    result = await session.execute(select(Doctors).where(Doctors.id == doctor_id))
    doctor = result.scalar_one_or_none()
    if not doctor:
        return {"message": "Doctor not found"}
    if doctor.role_id != 1:
        result = await session.execute(select(Patients).where(
            Patients.doctor_id == doctor_id, Patients.is_deleted == False))
    else:
        result = await session.execute(select(Patients).where(
            Patients.is_deleted == False))
    patients = result.scalars().all()
    return [patient.to_dict() for patient in patients]


@router.get("/get_last_upload_video_patient/{doctor_id}")
async def get_last_upload_video_patient(doctor_id: int, session: SessionDep = SessionDep):
    result = await session.execute(select(Doctors).where(Doctors.id == doctor_id))
    doctor = result.scalar_one_or_none()
    if not doctor:
        return {"message": "Doctor not found"}

    # Build the query using SQLAlchemy 2.0 syntax
    query = select(VideoPath).join(
        Patients, VideoPath.patient_id == Patients.id)

    if doctor.role_id != 1:  # Not an admin
        query = query.where(Patients.doctor_id == doctor_id)

    query = query.where(
        Patients.is_deleted == False,
        VideoPath.is_deleted == False,
        # REMOVED: VideoPath.original_video == True,
    )

    result = await session.execute(query)
    all_videos = result.scalars().all()

    if not all_videos:
        return {"message": "No video found", "patient_id": None, "video_id": None}

    # Step 2: Sort the results reliably in Python.
    latest_video = sorted(
        all_videos, key=lambda v: v.create_time, reverse=True)[0]

    return {
        "patient_id": latest_video.patient_id,
        "video_id": latest_video.id,
        "message": "success",
    }


@router.get("/check_patient_id_exists/{patient_id}")
async def check_patient_id_exists(patient_id: int, session: SessionDep = SessionDep):
    """检查病历号是否已存在"""
    result = await session.execute(select(Patients).where(
        Patients.id == patient_id, Patients.is_deleted == False))
    patient = result.scalar_one_or_none()
    return {"exists": patient is not None}


@router.put("/update_patient_by_id")
async def update_patient_by_id(patient: UpdatePatientModel = Body(..., embed=True), session: SessionDep = SessionDep):
    result = await session.execute(select(Doctors).where(
        Doctors.id == patient.doctor_id, Doctors.is_deleted == False))
    doctor = result.scalar_one_or_none()
    if not doctor:
        return {"message": "Doctor not found"}

    # 查找要更新的患者
    if doctor.role_id != 1:
        result = await session.execute(select(Patients).where(
            Patients.id == patient.patient_id,
            Patients.doctor_id == patient.doctor_id,
            Patients.is_deleted == False))
    else:
        result = await session.execute(select(Patients).where(
            Patients.id == patient.patient_id, Patients.is_deleted == False))
    patient_ = result.scalar_one_or_none()
    if not patient_:
        return {"message": "Patient not found or this doctor does not have permission to update this patient"}

    # 如果要修改病历号，需要特殊处理
    if patient.new_patient_id != patient.patient_id:
        # 检查新病历号是否已存在
        result = await session.execute(select(Patients).where(
            Patients.id == patient.new_patient_id, Patients.is_deleted == False))
        if result.scalar_one_or_none():
            return {"message": "New patient ID already exists"}

        # 创建新的患者记录
        new_patient = Patients(
            username=patient.username,
            age=patient.age,
            gender=patient.gender,
            case_id=patient.case_id,
            doctor_id=patient.doctor_id,
            create_time=patient_.create_time,  # 保持原创建时间
            update_time=datetime.now().strftime("%Y-%m-%d %H:%M:%S"),
            is_deleted=False
        )
        new_patient.id = patient.new_patient_id
        session.add(new_patient)

        # 更新相关表的外键引用

        # 更新VideoPath表
        await session.execute(
            update(VideoPath).where(
                VideoPath.patient_id == patient.patient_id
            ).values(patient_id=patient.new_patient_id)
        )

        # 更新Action表
        await session.execute(
            update(Action).where(
                Action.patient_id == patient.patient_id
            ).values(patient_id=patient.new_patient_id)
        )

        # 删除旧的患者记录
        patient_.is_deleted = True

        await session.commit()
        await session.refresh(new_patient)
        return new_patient.to_dict()
    else:
        # 如果不修改病历号，正常更新
        patient_.age = patient.age
        patient_.gender = patient.gender
        patient_.case_id = patient.case_id
        patient_.doctor_id = patient.doctor_id
        patient_.username = patient.username
        patient_.update_time = datetime.now().strftime("%Y-%m-%d %H:%M:%S")

        await session.commit()
        await session.refresh(patient_)
        return patient_.to_dict()


@router.delete("/delete_patient_by_id/{patient_id}/{doctor_id}")
async def delete_patient_by_id(patient_id: int, doctor_id: int, session: SessionDep = SessionDep):
    result = await session.execute(select(Doctors).where(
        Doctors.id == doctor_id, Doctors.is_deleted == False))
    doctor = result.scalar_one_or_none()
    if not doctor:
        return {"message": "Doctor not found"}
    if doctor.role_id != 1:
        result = await session.execute(select(Patients).where(
            Patients.id == patient_id, Patients.doctor_id == doctor_id, Patients.is_deleted == False))
        patient_ = result.scalar_one_or_none()
        if not patient_:
            return {"message": "Patient not found or this doctor does not have permission to delete this patient"}
    else:
        result = await session.execute(select(Patients).where(
            Patients.id == patient_id, Patients.is_deleted == False))
        patient_ = result.scalar_one_or_none()
        if not patient_:
            return {"message": "Patient not found"}
    patient_.is_deleted = True
    patient_.update_time = datetime.now().strftime("%Y-%m-%d %H:%M:%S")
    await session.commit()
    return {"message": "Patient deleted successfully"}


@router.get("/get_patients_with_page")
async def get_patient_with_page(
    page: int = Query(default=1, ge=1, description="页码"),
    page_size: int = Query(default=10, ge=1, le=100, description="每页数量"),
    sort_by: Optional[str] = Query(
        default="id", description="排序字段(id, username, email, create_time)"),
    sort_order: Optional[str] = Query(
        default="asc", description="排序方向(asc, desc)"),
    doctor_id: int = Query(None, ge=1),
    session: SessionDep = SessionDep
):

    # Build base query
    query = select(Patients).where(Patients.is_deleted == False)
    if doctor_id:
        query = query.where(Patients.doctor_id == doctor_id)

    # Apply sorting
    if sort_by:
        if sort_by == "id":
            query = query.order_by(Patients.id)
        elif sort_by == "username":
            query = query.order_by(Patients.username)
        elif sort_by == "create_time":
            query = query.order_by(Patients.create_time)

    # Get total count
    count_query = select(func.count(Patients.id)).where(
        Patients.is_deleted == False)
    if doctor_id:
        count_query = count_query.where(Patients.doctor_id == doctor_id)
    result = await session.execute(count_query)
    total = result.scalar()

    # Apply final sorting and pagination
    if sort_order == "desc":
        query = query.order_by(desc(Patients.id))
    else:
        query = query.order_by(asc(Patients.id))

    query = query.offset((page - 1) * page_size).limit(page_size)
    result = await session.execute(query)
    patients = result.scalars().all()

    return {
        "total": total,
        "page": page,
        "page_size": page_size,
        "patients": [patient.to_dict() for patient in patients]
    }
